### PR TITLE
State sync test

### LINKED ISSUE #7349: Add unit tests for state-sync
## What is the Problem Being Solved?

#7225 is likely to land with only integration tests. However there are some edge cases that the end to end test does not exercise.
To get coverage for those, we need some unit tests of at least the golang ExtensionSnapshotter logic.

## Description of the Design

Tests covering the following edge cases in golang:
- Reaching a snapshot block height when a snapshot is still in progress
- Reaching commit time with an active snapshot that hasn't returned from it's request to JS
  - Make sure commit is delayed through `WaitUntilSnapshotStarted` correctly stalling
- Reaching a second commit after snapshot has started: make sure no wait
- JS Snapshot initiate request fails
- JS Snapshot retrieval request fails

In an ideal world, we'd have tests covering the following edge cases in JS:
- Snapshot initiates request when a snapshot is already in progress
- Commit action before snapshot initiate completes (verify that swingstore commit is delayed)
- Snapshot discard before init completes
- Snapshot discard while snapshot in progress. Make sure any retrieval request is rejected
- Snapshot discard after snapshot data dir already removed

### PR DESCRIPTION
**Pull request by JimLarson**
closes: #7349

## Description

Adds the requested unit tests to the `SwingsetSnapthotter`. The stretch goal JS tests in #7349 were not reached, so leaving that issue open for now.

Refactored `SwingsetSnapshotter` for simpler test setup with simpler control points. The tests are not black-box and require access to unexported fields.

Lastly, added an explicit `RWMutex` to control reading/writing the `activeSnapshot` field that is accessed by multiple goroutines. Though there was a discipline of one goroutine only changing from nil to non-nil, and another goroutine from non-nil to nil, the discipline is undocumented and fragile. Best practice advises us to just use a more explicit and standard idiom. (@michaelfig please look closely at this in particular.)

### Security Considerations

None beyond the correctness.

### Scaling Considerations

The changes shouldn't be significant.

### Documentation Considerations

N/A

### Testing Considerations

Note that some of the tests spawn a goroutine outside of a WaitGroup, so there is no guarantee that the goroutine has exited when the test ends. However, we do wait for all visible effects of the goroutine, so it should not matter when it ends.

### CODE REVIEW

**Comment by mhofman** [comment]
> Lastly, added an explicit `RWMutex` to control reading/writing the `activeSnapshot` field that is accessed by multiple goroutines. Though there was a discipline of one goroutine only changing from nil to non-nil, and another goroutine from non-nil to nil, the discipline is undocumented and fragile. Best practice advises us to just use a more explicit and standard idiom. (@michaelfig please look closely at this in particular.)

I don't understand the need for that change. We already effectively had a mutex equivalent through the channel usage, and the order of operations, which was fully documented. I don't understand the costs of mutexes in Go, but I wanted to avoid incurring that cost in the first place since the vast majority of the time, there is no active snapshot.

**Comment by JimLarson** [comment]
The cost of taking a mutex lock should be negligible unless it's in a high-traffic inner loop. In this case it's running in some per-block code, which should be absolutely fine compared to all the signature validations, etc.

The visibility of updates to `activeSnapshot` isn't always mediated by channel or other synchronization. The nil'ing out of `activeSnapshot` on line 163 or 179 (before the PR changes) takes place after the channel message, so could be seen at any time, and could possibly even flip back to its old state on a subsequent read (maybe - I'd have to read the memory model very closely).

In fact, on some architectures the pointer update could be memory-unsafe since it doesn't use the `atomic` library.

An alternative fix would be to give the main goroutine exclusive control of the `activeSnapshot` field and have the worker goroutine signal completion by another channel or different use of the existing channel. I'm pretty sure we can change `SnapshotExtension()` to not worry about `activeSnapshot` being nil since it can't flip until that method returns.

File: golang/cosmos/x/swingset/keeper/snapshotter.go [hunk 1]
```diff
@@ -64,10 +65,19 @@ type SwingStoreExporter interface {
 }
 
 type SwingsetSnapshotter struct {
-	app            *baseapp.BaseApp
```
Comment (michaelfig) [inline golang/cosmos/x/swingset/keeper/snapshotter.go:67]: I appreciate the decoupling of the snapshotter from `baseApp` (presumably for ease of testing).

File: golang/cosmos/x/swingset/keeper/snapshotter.go [hunk 2]
```diff
@@ -64,10 +65,19 @@ type SwingStoreExporter interface {
...
+	logger            log.Logger
+	exporter          SwingStoreExporter
+	blockingSend      func(action vm.Jsonable) (string, error)
+
+	// activeSnapshot points to the active snapshot.
+	// Must hold the snapshotter's RWMutex to read or write it.
+	// The main goroutine moves it from nil to non-nil,
+	// and the goroutine spawned by InitiateSnapshot() moves it from
+	// non-nil back to nil.
 	activeSnapshot *activeSnapshot
```
Comment (michaelfig) [inline golang/cosmos/x/swingset/keeper/snapshotter.go]: I carefully examined the locking code, and while I struggled to stay awake looking at it, I *believe* it made sense. I find myself wishing for an abstraction like 
`activeSnapshot "sync/atomic".Pointer[activeSnapshot]`, especially since the RWLock is only there to protect loads and stores of the `activeSnapshot` pointer.

Since we're already on Go 1.19, I wouldn't fight against using generics.  Consider it?

**Review by michaelfig — APPROVED** [review]
LGTM, but please seriously consider relying on `"sync/atomic".Pointer[activeSnapshot]` to reduce the cognitive load of understanding this code.

File: golang/cosmos/x/swingset/keeper/snapshotter.go (hunk 2 above)
Comment (ivanlei) [reply golang/cosmos/x/swingset/keeper/snapshotter.go]: Totally agree. [sync.atomic.XXXPointer](https://pkg.go.dev/sync/atomic) is the way to go.

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		t.Error("wanted error for snapshot in progress")
+	}
+}
+
+func TestNotConfigured(t *testing.T) {
+	swingsetSnapshotter := newTestSnapshotter()
+	swingsetSnapshotter.isConfigured = func() bool { return false }
+	err := swingsetSnapshotter.InitiateSnapshot(123)
+	if err == nil {
+		t.Error("wanted error for unconfigured snapshot")
```
Comment (mhofman) [inline golang/cosmos/x/swingset/keeper/snapshotter_test.go]: nit
```suggestion
		t.Error("wanted error for unconfigured snapshot manager")
```

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = swingsetSnapshotter.InitiateSnapshot(123)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Second run through app.Commit() - should return right away
+	err = swingsetSnapshotter.WaitUntilSnapshotStarted()
```
Comment (mhofman) [inline golang/cosmos/x/swingset/keeper/snapshotter_test.go:80]: Ugh? This is where it should block. Maybe I misphrased the issue:
- The first commit initiates a snapshot
- The second commit blocks until the snapshot has started
- (optional) the 3rd commit proceed without blocking even if the snapshot is still in progress.

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	err := swingsetSnapshotter.InitiateSnapshot(123)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = swingsetSnapshotter.WaitUntilSnapshotStarted()
+	if err == nil {
+		t.Fatal("wanted initiation error")
+	}
```
Comment (mhofman) [inline golang/cosmos/x/swingset/keeper/snapshotter_test.go:106]: Maybe we should make clear in the test that subsequent calls to `WaitUntilSnapshotStarted` will not report the error.

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go
Comment (mhofman) [inline golang/cosmos/x/swingset/keeper/snapshotter_test.go:1]: I just realized another path we should test:
- If the cosmos `Snapshot` doesn't call `SnapshotExtension`, we correctly send a discard message to the controller.

**Review by mhofman — CHANGES_REQUESTED** [review]
Thanks for the refactors.

I did not realize pointer read/writes were not guaranteed to be atomic. I do understand the need for being explicit there, but as @michaelfig , the mutex dance seem cognitively onerous (for example there is a case where the write lock is taken even though we're only reading the pointer, but assigning a value in the structure pointed at).

Regarding the test itself, I'm unsure where the disconnect is, but it's the `WaitUntilSnapshotStarted` immediately following an `InitiateSnapshot` which should block. I do remember manually testing this when developing the feature and it was working as expected. I'm still struggling to read channel blocking logic in golang, so maybe the problem is in the test construction somehow?

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go (hunk 3 above)
Comment (JimLarson) [reply golang/cosmos/x/swingset/keeper/snapshotter_test.go]: Done.

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go
Comment (JimLarson) [reply golang/cosmos/x/swingset/keeper/snapshotter_test.go:1]: There's a synchronization problem for testing in that the discard message could come at any time - the test doesn't know how long to wait for it without some other synchronization mechanism. But we might want just such a mechanism instead of mutexes or atomics. I'll chat with you to discuss.

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go (hunk 5 above)
Comment (JimLarson) [reply golang/cosmos/x/swingset/keeper/snapshotter_test.go:106]: Added another call and it works.

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go (hunk 4 above)
Comment (JimLarson) [reply golang/cosmos/x/swingset/keeper/snapshotter_test.go:80]: I made a note in #7349 that it's not practical to write tests that ensure that a goroutine is blocked - there's no "run until quiescent" command for the scheduler. (We asked Rob Pike for this in the early days, having had experience with nightly pseudorandom distributed system simulations tests for our Paxos system. Our request was denied.) The alternative is to put in a pause so that it's likely to be blocked if it hadn't finished - but that's flaky and makes the tests unnecessarily long.

So this test is meant to cover the 3rd bullet item from #7349, not the 2nd.

The test here allows the first controller call to complete immediately, which is why line 62 doesn't block. (The channel is created with a buffer of 1 on line 45).

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go
Comment (mhofman) [reply golang/cosmos/x/swingset/keeper/snapshotter_test.go:1]: Can't we have `blockingSend` use a channel to signal when it sees the discard message? Or is the problem that if it fails, the test will deadlock (i.e. is there a deadlock detection mechanism in the test runner to report such test cases as failed?)

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go (hunk 4 above)
Comment (mhofman) [reply golang/cosmos/x/swingset/keeper/snapshotter_test.go:80]: Oh yeah I definitely don't want flaky time based tests.

Can we at the very least test the internal structure used for the synchronization? Would the following work:

- call `InitiateSnapshot`
- verify that `activeSnapshotter` is not nil
- write in channel to unblock goroutine waiting on blockingSend
- call `WaitUntilSnapshotStarted`
- verify that `activeSnapshotter` is nil

File: golang/cosmos/x/swingset/keeper/snapshotter.go (hunk 2 above)
Comment (JimLarson) [reply golang/cosmos/x/swingset/keeper/snapshotter.go]: I took a look and the `"sync/atomic" package right away echoes the advice at the top of the [Go memory model](https://go.dev/ref/mem): if you're using this, you're probably doing something wrong. It then repeats the Go slogan "Share memory by communicating; don't communicate by sharing memory."

Inspired by this I tried using a more idiomatic "done" channel for the worker goroutine to signal that it had finished, letting the main thread be in sole control of updates to the `activeSnapshot` field. It actually leads to simplifications in the code and allows more expressive tests. PTAL.

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go
Comment (JimLarson) [reply golang/cosmos/x/swingset/keeper/snapshotter_test.go:1]: I used a "done" channel for the worker goroutine to signal its completion, which solves this problem. PTAL

File: golang/cosmos/x/swingset/keeper/snapshotter_test.go (hunk 4 above)
Comment (JimLarson) [reply golang/cosmos/x/swingset/keeper/snapshotter_test.go:80]: I think I've implemented what you want a slightly different way.

**Review by michaelfig — APPROVED** [review]
LGTM!  Much tidier with channels. :)

**Review by mhofman — APPROVED** [review]
The `done` channel change looks good. From what I gather it's very close to my original implementation, except that now only the main thread assigns the `activeSnapshot` pointer by requiring `WaitUntilSnapshotStarted` and `InitiateSnapshot` do both cleanup a previously completed snapshot.

Still unfortunate that we don't get to test the blocking behavior of `WaitUntilSnapshotStarted`.
